polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -273,25 +276,42 @@ class IOSharedNotebook(
       }
     }
 
-    def runCell(id: CellID): IO[Stream[IO, Result]] = withInterpreterLaunch(id) {
-      kernel =>
-        val buf = new WindowBuffer[Result](1000)
-        kernel.runCell(id).map {
-          results =>
-            results.evalTap {
-              // buffer the result and also broadcast to all clients
-              result => IO(buf.add(result)) *> outputMessages.publish1(CellResult(ShortString(path), id, result))
-            }.evalTap {
-              // if there are any UpdatingDataReprs, watch for their updates and broadcast
-              case v: ResultValue => watchUpdatingValues(v).through(outputMessages.publish).compile.drain.start.as(()) // TODO: is it wise to forget the fiber?
-              case _ => IO.unit
-            }.onFinalize {
-              // write the buffered results to the notebook
-              ref.update {
-                case (ver, nb) => ver -> nb.setResults(id, buf.toList)
-              }
+    override def runCell(id: CellID): IO[Stream[IO, Result]] = queueCell(id).flatten
+
+    def queueCell(id: CellID): IO[IO[Stream[IO, Result]]] = get.flatMap {
+      notebook =>
+        notebook.getCell(id).filterNot(_.language == "text").fold[IO[IO[Stream[IO, Result]]]](IO.pure(IO.pure(Stream.empty))) {
+          _ =>
+            withInterpreterLaunch(id) {
+              kernel =>
+                val buf = new WindowBuffer[Result](1000)
+                kernel.queueCell(id).map {
+                  ioResult => ioResult.map {
+                    results =>
+                      results.evalTap {
+                        // buffer the result and also broadcast to all clients
+                        result => IO(buf.add(result)).flatMap {
+                          _ =>
+                            outputMessages.publish1(CellResult(ShortString(path), id, result))
+                        }
+                      }.evalTap {
+                        // if there are any UpdatingDataReprs, watch for their updates and broadcast
+                        case v: ResultValue => watchUpdatingValues(v).through(outputMessages.publish).compile.drain.start.as(()) // TODO: is it wise to forget the fiber?
+                        case _ => IO.unit
+                      }.onFinalize {
+                        // write the buffered results to the notebook
+                        ref.update {
+                          case (ver, nb) => ver -> nb.setResults(id, buf.toList)
+                        }
+                      }
+                  }.handleErrorWith {
+                    case errs@CompileErrors(_) => IO.pure(Stream.emit(errs))
+                    case err@RuntimeError(_) => IO.pure(Stream.emit(err))
+                    case err => IO.pure(Stream.emit(RuntimeError(err)))
+                  }
```

Comment:
Hmm, do we need this error handling both here and in `PolyKernel.queue`?

Reply:
Probably not? 🤷‍♂️